### PR TITLE
[CIT-659] restart device local operation

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

We can now restart the device by publishing on the topic `tedge/commands/req/control/restart` with the id of the operation.

1. The previous code was reused as much as possible. Structs like ~~`SoftwareOperationStatus`~~ became `OperationStatus`. 
2. New request/response topics were created for this operation.
3. ~~`fail_pending_operations`~~ became `process_pending_operation`. This is the function that is executed when we start `tedge_agent`. This function now verifies if the restart was successful - making the rename appropriate. 
4. `StateStatus` is the struct that holds all possible operations `{Software, Restart, UnknownOperation}`. Where, `Software` made up of the __enum__ `SoftwareOperationVariants::{List, Update}` and `Restart` is made up of the __enum__ `RestartOperationStatus::{Pending, Restarting}`.
5. The restart operation functions are separated as a module. 

The restart logic goes as follows: 
1. receive restart message etc...
2. set operation status to `RestartOperationStatus::Pending`
3. create `/run/tedge_agent_restart` with current unix timestamp (in UTC)
4.  set operation status to `RestartOperationStatus::Restarting`
3.  Call `sync` command __then__ call `init 6` to reboot.
4. after reboot, the agent restarts calling the `process_pending_operation` where we __check__ if the reboot was a success.
5. if `/run/tedge_agent_restart` does __NOT__ exist, we conclude this was removed by the filesystem on reboot
6. if `/run/tede_agent_restart` __exists__, we compare the first value of `/proc/uptime` (seconds the system has been on) with the unix timestamp inside `/run/tedge_agent_restart` to decide if the restart operation was successful. 

## How to test
NOTE: this will not restart your device. The `init 6` command is currently changed to `echo 6` :)
After building and stopping existing agent (you might need to remove the lockfile in /run/lock/):
```
sudo ./target/debug/tedge_agent
```

```
mosquitto_sub -t '#' -v
```

send a restart operation:
```
mosquitto_pub -t 'tedge/commands/req/control/restart' -m '{"id":"1234"}'
```

Stop ./target/debug/tedge_agent and run again. Bullet points 4,5,6 (in the section above) detail the restart check, but to summarise again: 

1. if your computer has only been turned on in the last 5 minutes the reboot will pass (MAX_ALLOWED_DURATION in the code).
2. if you manually delete the /run/tedge_agent_restart file before running the `tedge_agent` component, the reboot will pass 

## Out of scope of this ticket
Also, I took the liberty to remove `chrono` as a dependency completely from the `tedge_agent` component - I hope you do not mind. It was a minor change, only in one file :) 

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://cumulocity.atlassian.net/browse/CIT-659

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by initard** [comment]
testing

File: sm/tedge_agent/src/restart_operation_handler.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
+pub mod restart_operation {
+
+    use crate::error::AgentError;
+    use std::{fs::File, fs::OpenOptions, io::Read, io::Write, path::Path};
+    use time::OffsetDateTime;
+    use tracing::error;
+
+    const SLASH_RUN_PATH_TEDGE_AGENT_RESTART: &str = "/run/tedge_agent_restart";
```
Comment (rina23q) [inline sm/tedge_agent/src/restart_operation_handler.rs:8]: I think better to create a directory owned by `tegde-agent` during the installation script, and create a file under the directory. Then, the `tedge_agent` run by `tedge-agent` user can create a run file. For example, `/run/tedge_agent/restart`.

**Review by rina23q — COMMENTED** [review]
Left some comments.

File: sm/tedge_agent/src/restart_operation_handler.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                    let local = OffsetDateTime::now_utc();
+                    let uptime_secs = result as i64;
+                    let duration = time::Duration::seconds(uptime_secs);
+                    let reboot_time = local - duration;
+                    Ok(reboot_time)
+                }
+                Err(_err) => Err(AgentError::FloatCastingError),
+            }
+        } else {
+            return Err(AgentError::UptimeParserError);
```
[the remark this answers is not in the document]
Comment (initard) [reply sm/tedge_agent/src/restart_operation_handler.rs:83]: Thank you, I see it is not needed :)

File: sm/tedge_agent/src/agent.rs [hunk 3]
```diff
@@ -44,7 +50,8 @@ impl Default for SmAgentConfig {
 
         let mqtt_client_config = mqtt_client::Config::default().with_packet_size(10 * 1024 * 1024);
 
-        let request_topics = TopicFilter::new(software_filter_topic()).expect("Invalid topic");
+        let mut request_topics = TopicFilter::new(software_filter_topic()).expect("Invalid topic");
+        let () = request_topics.add(control_filter_topic()).unwrap(); // TODO: fix this unwrap
```
[the remark this answers is not in the document]
Comment (initard) [reply sm/tedge_agent/src/agent.rs:54]: Yes, done!

File: sm/tedge_agent/src/restart_operation_handler.rs (hunk 1 above)
Comment (initard) [reply sm/tedge_agent/src/restart_operation_handler.rs:8]: But on almost all systems `/run` is cleared on system restart @rina23q

File: sm/tedge_agent/src/agent.rs [hunk 4]
```diff
@@ -345,29 +367,130 @@ impl SmAgent {
         Ok(())
     }
 
-    async fn fail_pending_operation(&self, mqtt: &Client) -> Result<(), AgentError> {
+    async fn handle_restart_operation(
```
Comment (didier-wenzek) [inline sm/tedge_agent/src/agent.rs:370]: This function has far too many responsibilities.

* Decoding the request message
* Managing the agent state
* Forwarding the operation status to the requester
* And doing the restart.

This should be at least be broken in more specific functions. More importantly error handling must be done *around* all these steps to ensure that any error is properly handled.

File: sm/tedge_agent/src/agent.rs [hunk 5]
```diff
@@ -345,29 +367,130 @@ impl SmAgent {
...
+                }
+                .into());
+            }
+        };
+
+        self.persistance_store
+            .update(&StateStatus::Restart(RestartOperationStatus::Restarting))
+            .await?;
+
+        let () = restart_operation::create_slash_run_file()?;
```
Comment (didier-wenzek) [inline sm/tedge_agent/src/agent.rs:407]: An error here will not be neither reported to the caller nor stored in the persistent state.

* This is the case for most of the errors raised in the `handle_restart_operation`.
* A `?` is akin to throwing an exception. One need a place to catch these errors.

**Review by didier-wenzek — COMMENTED** [review]
The `handle_restart_operation` function deserves a refactoring.

Another refactoring would be to rethink the way messages are decoded in the main `subscribe_and_process` loop. The responsibility to decode a message depending of the topic should be moved to the `sm/json_sm/src/messages.rs` module. This is more involved and a bit out scope though.

File: sm/tedge_agent/src/agent.rs [hunk 6]
```diff
@@ -31,8 +35,10 @@ pub struct SmAgentConfig {
     pub request_topic_list: Topic,
     pub request_topic_update: Topic,
     pub request_topics: TopicFilter,
+    pub request_topic_restart: Topic,
     pub response_topic_list: Topic,
     pub response_topic_update: Topic,
+    pub response_topic_restart: Topic,
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply sm/tedge_agent/src/agent.rs:47]: Ok we will see later.

File: sm/tedge_agent/src/agent.rs [hunk 7]
```diff
@@ -397,20 +520,48 @@ async fn publish_capabilities(mqtt: &Client) -> Result<(), AgentError> {
...
+    about = clap::crate_description!()
+    )]
+    pub struct AgentOpt {
+        /// Turn-on the debug log level.
+        ///
+        /// If off only reports ERROR, WARN, and INFO
+        /// If on also reports DEBUG and TRACE
+        #[structopt(long)]
+        pub debug: bool,
+    }
```
Comment (didier-wenzek) [inline sm/tedge_agent/src/agent.rs:542]: I don't understand the purpose of this struct.

**Review by didier-wenzek — COMMENTED** [review]
Just some minor issues to be fixed.

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**